Custom Labels & Date/Time Dimensions

natefaubion opened the pull request.

natefaubion commented:
@cryogenian This is ready for the pivot table, at least. The chart configs can be done next. The most relevant portions for you would be `Setups.Dimension` and the stuff I added to `Setups.Inputs`.

jdegoes commented:
@cryogenian Count should be there.

natefaubion commented:
@cryogenian All those are part of the Transform stuff (there's an Aggregation transform). When you click the configure button, it brings up the action selector. https://github.com/slamdata/slamdata/issues/1491#issuecomment-285490842

I did not add `COUNT` as a transform yet. Currently pivot table only allows you to select `COUNT(*)` as a category, but adding it as a transform is straightforward. Since there will be other transforms (per #1491), I figured we could add those in another PR. No functionality is lost in this PR.

jdegoes commented:
If the user could select root (`*`) in the dimension picker (which of course is not possible now), then the user could select root and choose to count it.

natefaubion commented:
That was something I considered when doing the PR for selecting raw values. I didn't really want to always have a column for the root value, and require users select it. But there's no reason I couldn't provide `*` as an option in place of `COUNT`, and keep it as a leaf value.

natefaubion commented:
Right now the dimension picker only supports leaf values though, so I don't know what it would mean to tabulate `*`. There's currently no way for it to display nested records, which is why `COUNT(*)` is a special case.

jdegoes commented:
@natefaubion Sure, but we can enable / disable verbs based on the data type. And most likely the pivot table limitation of only displaying leaf values is a temporary restriction until such point as we have a sane jtable implementation.

jdegoes commented:
Btw, the reason I bring this up is because in @garyb's structure editor work, the root is exposed so you can manipulate the whole set.

cryogenian commented:
> Is column header for category axes supposed to have [Min, Max, Sum, Average] options?
@natefaubion I mean there is not much sense in `Average` when your column contains strings only. 
Could you please explain how can I select `COUNT` as dimension or column?

natefaubion commented:
> Sure, but we can enable / disable verbs based on the data type. And most likely the pivot table limitation of only displaying leaf values is a temporary restriction until such point as we have a sane jtable implementation.

I just mean, right now it only selects leaf values, so I would have to expose `*` as a leaf value. If someone selects `*` and applies no verbs to it, then I'm not sure what I would display except a blank cell.

> @natefaubion I mean there is not much sense in Average when your column contains strings only.

You are right. It uses the same verbs for all the columns, as that was the previous behavior. 

> Could you please explain how can I select COUNT as dimension or column?

When you add a column, it's the first item listed in the picker.

---

I figured I could do the rest of the transform work in a separate PR, as I still have design questions around some of them, but I can just continue on this one if you'd rather.

jdegoes commented:
If there is only one verb choice that makes sense, we should probably auto-select it.

cryogenian commented:
@natefaubion Hm... I can't see `COUNT` probably because of `unfoldColumn + groupColumn`. `Count` is unfolded to `Nil` -> `L.zip (pure root) $ Nil == Nil` -> `CF.unfoldCofree` ignores `Count`.

cryogenian commented:
@natefaubion I'm starting to work on this branch. Will make prs to it

natefaubion commented:
@cryogenian Hmm, you are right. @garyb any ideas why `Count` might be omitted from the dimension list?

natefaubion commented:
`Count` yields `Nothing` because it's a leaf node, but I would expect it to still be in the list.

cryogenian commented:
@natefaubion I've made a pr to this branch where `Count -> Just $ Tuple Count $ Column JCursorTop` that works

garyb commented:
I thought we already had a fix in for that, but I guess not for this case...

garyb commented:
``` purescript
groupColumns ∷ L.List Column → CF.Cofree L.List ColumnNode
groupColumns cs =
  let
    root = Column { value: J.JCursorTop, valueAggregation: Nothing }
    tree = constructTree unfoldColumn root cs
  in
    discriminateNodes
      if extract <$> CF.tail tree == pure Count -- <---- this `if` is supposed to be the fix
      then CF.mkCofree root (CF.mkCofree Count L.Nil : CF.mkCofree root L.Nil : L.Nil)
      else tree
```

I originally proposed that as `L.null (CF.tail tree)` but I think you said that wouldn't work because `COUNT` is always present @natefaubion - perhaps it's the reverse though, and the null check would be another way of fixing this.

natefaubion commented:
@garyb the problem is the opposite. When there are a bunch of fields, COUNT doesn't show up at all :P.

garyb commented:
I knew that, I was just testing you :eyes:

natefaubion commented:
@cryogenian's PR fixes is though. Seems acceptable to me, at least.

natefaubion commented:
I've changed the `COUNT` column to `*` and made `COUNT` a transform (so now you can count anything). By default, unless `*` is a raw value, selecting `*` will automatically select the `COUNT` transform. I also added lower, upper, and length as string transforms. I think I will defer round/ceil/floor to a different ticket because they require quite a but more machinery, at both the UI level and codegen level (since I have to insert special functions in the query to simulate them).

natefaubion commented:
I've also made it so the columns infer options based on the axis type, so it doesn't always just present aggregations.

natefaubion commented:
<img width="674" alt="screen shot 2017-03-14 at 9 33 59 am" src="https://cloud.githubusercontent.com/assets/144058/23905680/d0cb989a-0899-11e7-85f9-4b92cac2c94d.png">
